File: src/AnalyticsService/Service/Dto/CloudEvent.fs
namespace AnalyticsService.Service.Dto

open System
open System.Text.Json.Serialization

/// A record representing a cloud event from MQ.
type CloudEvent<'T> = {
    [<JsonPropertyName("id")>]
    Id: Guid

    [<JsonPropertyName("data")>]
    Data: 'T

    [<JsonPropertyName("datacontenttype")>]
    ContentType: string

    [<JsonPropertyName("source")>]
    Source: string
}

File: src/AnalyticsService/Service/Model/CloudEvent.cs
using System.Text.Json.Serialization;

namespace AnalyticsService.Service.Model;

/// <summary>
/// A record encapsulating a cloud event.
/// </summary>
internal sealed record CloudEvent<T>(
    [property: JsonPropertyName("source")]
    string Source,
    [property: JsonPropertyName("topic")]
    string Topic,
    [property: JsonPropertyName("data")]
    T Data
);